src/iop/opposed.c

```diff
@@ -71,8 +71,7 @@ static float * _process_linear_opposed(dt_dev_pixelpipe_iop_t *piece, const void
   const size_t p_size = (size_t) dt_round_size(pwidth * pheight, 16);
 
   int *mask_buffer = dt_calloc_align(64, 4 * p_size * sizeof(int));
-  const size_t bsize = (4 + MAX(roi_in->width + roi_in->x, roi_out->width + roi_out->x)) * (4 + MAX(roi_in->height + roi_in->y, roi_out->height + roi_out->y)); 
```

Comment:
I don't understand the `roi_in->width + roi_in->x`. Here x is the starting point and width the size. So for the buffer we indeed just need width as done below now. Now for the `8 *` don't know...

Reply:
Should this roi_out in this line:
https://github.com/darktable-org/darktable/blob/b51a81c65eed677d38d5cdfc5bfe8cb20355bebd/src/iop/opposed.c#L107

Reply:
Same question here:

https://github.com/darktable-org/darktable/blob/b51a81c65eed677d38d5cdfc5bfe8cb20355bebd/src/iop/opposed.c#L181

Reply:
And here:

https://github.com/darktable-org/darktable/blob/b51a81c65eed677d38d5cdfc5bfe8cb20355bebd/src/iop/opposed.c#L266

Reply:
Well, the tmp buffer is holding intermediate data corresponding to roi_in data. See the alloc.

Reply:
Ok, I see... Reading the code it seems that only a `4 *` should be needed for tmpout and not `8 *`.

Reply:
I am pretty sure the code is not here, even immediately going to error after the first loop leads to crashes.

Reply:
Both refavg functions are only called with row & col missing the outermost lines/rows.

Two things to mention

1. Even if I goto error after the first loop that just copies data I see crashes. 
2. I could only see crashes for xtrans! As the internal code is the same, is there some other code behaving bad changing roi_out size? Is there a way to check the size of the output buffer referred by the roi?

Reply:
I see, that's a strange issue.

To me the following line in the first loop looks strange:

`float *in = (float *)ivoid + (size_t)roi_in->width * (row + roi_out->y) * 4 + roi_out->x * 4;`

I'm not sure how in could be calculated using roi_out->y here and be sure that we don't go outside of ivoid buffer. But again I'm no exptert on this.

Maybe you could check if you still reproduce the crash but replacing:

`out[c] = fmaxf(0.0f, in[c]);`

with:

`out[c] = 1.0;`

That's really the only thing that could be wrong in this loop. But you probably already have tested this.

Reply:
Probably worth adding that into another debug option (roi) as it would be very helpful to track down such issues. TIA.

Reply:
Ok did that and already some interesting problem detected. In some cases the roi_out is larger than roi_in which comes at a surprise and definitely must lead to crashes with current code.

Reply:
> In some cases the roi_out is larger than roi_in

In this module or changed somewhere else? Note that a larger roi_out is expected in module doing distortion like perspective correction IINM.

Reply:
The roi scales in this module are both 1. If a module requests a larger roi_in area like for rotating this info is pushed to all modules earlier in the pipeline. No?

Who increases the roi? Not sure yet, anyway it must not be larger than what we get from rawprepare?

Reply:
Not sure but I think `borders.c` can increase the roi_out.

Reply:
I think I got it. Other modules (demosaic an example) can extend the roi because of rounding. That lead to pre-demosaic modules extending to but always staying in safe areas respecting demosaic patterns too. 

Here we extend the roi in, roi out needs a sanity check too.